zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/internal/BulkCallBuilder.java

```diff
@@ -148,13 +151,23 @@ public BulkCallBuilder(ElasticsearchStorage es, float esVersion, String tag) {
       return headers;
     }
 
-    @Override public void writeBody(HttpCall.RequestStream requestStream) {
-      for (IndexEntry<?> entry : entries) {
-        if (!requestStream.tryWrite(HttpData.wrap(serialize(alloc, entry, shouldAddType)))) {
-          // Stream aborted, no need to serialize anymore.
-          return;
-        }
+    @Override
+    public HttpRequest get() {
+      final HttpRequestWriter writer = HttpRequest.streaming(headers);
+      writeEntry(writer, 0);
+      return writer;
+    }
+
+    private void writeEntry(HttpRequestWriter writer, int index) {
+      if (index == entries.size()) {
+        writer.close();
+        return;
+      }
+      if (!writer.tryWrite(HttpData.wrap(serialize(alloc, entries.get(index), shouldAddType)))) {
```

Comment:
This looks fine but still curious why we need to use whenConsumed even though we have this boolean check here? I guess it's effectively respecting backpressure which is a good thing, but surprised it affects the stability.

Reply:
> but surprised it affects the stability.

This doesn't affect the stability, it's just for not buffering all the entries in advance before sending the request.
I think all the flaky tests were about `AggregatedRequestSupplier`, not `BulkCallBuilder`.

Previously, we are doing:
```java
HttpRequestWriter httpRequest = HttpRequest.streaming(request.headers());
// Send the request header to the server before we call httpRequest.close().
response = httpClient.execute(httpRequest);
request.writeBody(httpRequest::tryWrite);
httpRequest.close();
```
There's a high chance that the response is received before the request is complete.
So I changed to:
```java
HttpRequestWriter httpRequest = HttpRequest.streaming(request.headers());
request.writeBody(httpRequest::tryWrite);
httpRequest.close();
// Send the request header and body together to the server.
response = httpClient.execute(httpRequest);
```
which there's less chance that the response is received before the request is complete.

But as you noticed, this can be a problem for `BulkCallBuilder` when it's sending streaming requests.
So I changed to use backpressure for that so it's not buffering at all.

Reply:
A discussion like this begs for a comment, ya?

Reply:
Maybe this?

// There's a high chance that the response is received before the request is complete.
// This can be a problem for BulkCallBuilder when it's sending streaming requests.
// Hence, we use backpressure, so it's not buffering at all.

Reply:
but let me pass the build first. 😅

Reply:
Thanks for working on this - as an aside, if there isn't one already it'd probably be great to have a documentation setting on armeria.dev on how to do a streaming HTTP request :)

Reply:
Yeah, we need that. 😆

Reply:
The build passes. I didn't revert this change because I like this style but if you want the older version then I can revert it. 😄

Reply:
changed the comments per above I hope.